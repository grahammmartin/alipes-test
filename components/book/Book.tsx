"use client";
import {
  Menu,
  MenuButton,
  MenuItem,
  MenuItems,
  Transition,
} from "@headlessui/react";
import {
  EllipsisVerticalIcon,
  EyeIcon,
  PlusCircleIcon,
} from "@heroicons/react/20/solid";
import { BookType } from "@/app/books/page";

function classNames(...classes: any) {
  return classes.filter(Boolean).join(" ");
}

const Book = ({ title, author, price, description }: BookType) => {
  return (
    <div className="bg-white px-4 py-5 sm:px-6 border rounded-md w-64 h-64">
      <div className="flex space-x-3">
        <div className="min-w-0 flex-1 space-y-4">
          <h1 className="hover:underline">
            <p className="text-sm font-semibold text-gray-900">{title}</p>
          </h1>
          <p className="text-sm text-gray-500">Author : {author}</p>
          <p className="text-sm text-gray-500">price : {price}$</p>
          <p className="text-sm text-gray-500">description : {description}</p>
        </div>
        <div className="flex flex-shrink-0 self-start">
          <Menu as="div" className="relative inline-block text-left">
            <div>
              <MenuButton className="-m-2 flex items-center rounded-full p-2 text-gray-400 hover:text-gray-600">
                <span className="sr-only">Open options</span>
                <EllipsisVerticalIcon className="h-5 w-5" aria-hidden="true" />
              </MenuButton>
            </div>

            <Transition
              enter="transition ease-out duration-100"
              enterFrom="transform opacity-0 scale-95"
              enterTo="transform opacity-100 scale-100"
              leave="transition ease-in duration-75"
              leaveFrom="transform opacity-100 scale-100"
              leaveTo="transform opacity-0 scale-95"
            >
              <MenuItems className="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none">
                <div className="py-1">
                  <MenuItem>
                    {({ focus }) => (
                      <h3
                        className={classNames(
                          focus ? "bg-gray-100 text-gray-900" : "text-gray-700",
                          "flex px-4 py-2 text-sm"
                        )}
                      >
                        <EyeIcon
                          className="mr-3 h-5 w-5 text-gray-400"
                          aria-hidden="true"
                        />
                        <span>View Details</span>
                      </h3>
                    )}
                  </MenuItem>
                  <MenuItem>
                    {({ focus }) => (
                      <h3
                        className={classNames(
                          focus ? "bg-gray-100 text-gray-900" : "text-gray-700",
                          "flex px-4 py-2 text-sm"
                        )}
                      >
                        <PlusCircleIcon
                          className="mr-3 h-5 w-5 text-gray-400"
                          aria-hidden="true"
                        />
                        <span>Purchase</span>
                      </h3>
                    )}
                  </MenuItem>
                </div>
              </MenuItems>
            </Transition>
          </Menu>
        </div>
      </div>
    </div>
  );
};

export default Book;
