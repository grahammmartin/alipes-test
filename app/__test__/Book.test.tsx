import {
  fireEvent,
  render,
  screen,
  act,
  waitFor,
} from "@testing-library/react";
import Book from "@/components/book/Book";

/* eslint-disable */
class ResizeObserverMock {
  constructor() {}

  observe() {}

  unobserve() {}

  disconnect() {}
}
/* eslint-disable */
global.ResizeObserver = ResizeObserverMock;

describe("Book component", () => {
  it("should render the component correctly", () => {
    const { container } = render(
      <Book
        title="Test Book"
        author="Test Author"
        price={10}
        description="Test description"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders with correct title", () => {
    const { getByText } = render(
      <Book
        title="Test Book"
        author="Test Author"
        price={10}
        description="Test description"
      />
    );
    expect(getByText("Test Book")).toBeInTheDocument();
  });

  it("renders with correct author", () => {
    const { getByText } = render(
      <Book
        title="Test Book"
        author="Test Author"
        price={10}
        description="Test description"
      />
    );
    expect(getByText("Author : Test Author")).toBeInTheDocument();
  });

  it("renders with correct price", () => {
    const { getByText } = render(
      <Book
        title="Test Book"
        author="Test Author"
        price={10}
        description="Test description"
      />
    );
    expect(getByText("price : 10$")).toBeInTheDocument();
  });

  it("renders with correct description", () => {
    const { getByText } = render(
      <Book
        title="Test Book"
        author="Test Author"
        price={10}
        description="Test description"
      />
    );
    expect(getByText("description : Test description")).toBeInTheDocument();
  });

  it("renders with menu items", async () => {
    render(
      <Book
        title="Test Book"
        author="Test Author"
        price={10}
        description="Test description"
      />
    );
    const menu = screen.getByTestId("menu-box");
    fireEvent.click(menu);

    await waitFor(() => {
      const option1 = screen.getByText("View Details");
      const option2 = screen.getByText("Purchase");
      expect(option1).toBeInTheDocument();
      expect(option2).toBeInTheDocument();
    });
  });
});
