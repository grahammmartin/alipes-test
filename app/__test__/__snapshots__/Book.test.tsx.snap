// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book component should render the component correctly 1`] = `
<div>
  <div
    class="bg-white px-4 py-5 sm:px-6 border rounded-md w-64 h-64"
  >
    <div
      class="flex space-x-3"
    >
      <div
        class="min-w-0 flex-1 space-y-4"
      >
        <h1
          class="hover:underline"
        >
          <p
            class="text-sm font-semibold text-gray-900"
          >
            Test Book
          </p>
        </h1>
        <p
          class="text-sm text-gray-500"
        >
          Author : 
          Test Author
        </p>
        <p
          class="text-sm text-gray-500"
        >
          price : 
          10
          $
        </p>
        <p
          class="text-sm text-gray-500"
        >
          description : 
          Test description
        </p>
      </div>
      <div
        class="flex flex-shrink-0 self-start"
      >
        <div
          class="relative inline-block text-left"
          data-headlessui-state=""
        >
          <div>
            <button
              aria-expanded="false"
              aria-haspopup="menu"
              class="-m-2 flex items-center rounded-full p-2 text-gray-400 hover:text-gray-600"
              data-headlessui-state=""
              data-testid="menu-box"
              id="headlessui-menu-button-:r1:"
              type="button"
            >
              <span
                class="sr-only"
              >
                Open options
              </span>
              <svg
                aria-hidden="true"
                class="h-5 w-5"
                data-slot="icon"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 3a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM10 8.5a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM11.5 15.5a1.5 1.5 0 1 0-3 0 1.5 1.5 0 0 0 3 0Z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
